collections/src/main/scala/strawman/collection/immutable/LazyList.scala

```diff
@@ -220,6 +221,112 @@ sealed abstract class LazyList[+A] extends LinearSeq[A] with LazyListOps[A, Lazy
     if (this.isEmpty) z
     else tail.foldLeft(op(z, head))(op)
   }
+
+  /** Write all defined elements of this iterable into given string builder.
+    *  The written text begins with the string `start` and is finished by the string
+    *  `end`. Inside, the string representations of defined elements (w.r.t.
+    *  the method `toString()`) are separated by the string `sep`. The method will
+    *  not force evaluation of undefined elements. A tail of such elements will be
+    * represented by a `"?"` instead.  A cyclic stream is represented by a `"..."`
+    * at the point where the cycle repeats.
+    *
+    * @param b The [[collection.mutable.StringBuilder]] factory to which we need
+    * to add the string elements.
+    * @param start The prefix of the resulting string (e.g. "LazyList(")
+    * @param sep The separator between elements of the resulting string (e.g. ",")
+    * @param end The end of the resulting string (e.g. ")")
+    * @return The original [[collection.mutable.StringBuilder]] containing the
+    * resulting string.
+    */
+  protected def toStringBuilder(b: StringBuilder, start: String, sep: String, end: String): StringBuilder = {
+    b append start
+    if (!isEmpty) {
+      if (headDefined && !tailDefined) b append head // For LazyList, head is also Lazy. Hence adds `?` if it not defined yet
```

Comment:
Does `tailDefined` mean something in these 4 conditions? It seems it could be simplified to 2 cases: `headDefined` and `!headDefined`.

Reply:
I added those special cases and early returning `if` conditions to handle few special cases. For example, if `LazyList(1, 2, 3, 4)` is not evaluted then its `toString` would be `LazyList(?, ?)` if we don't handle `!headDefined and !tailDefined`.  Which is also more descriptive as it tells that both `head` and `tail` are not evaluated, but @julienrf  suggested that it should be just `LazyList(?)`.  See [this unit test](https://github.com/dhanesharole/collection-strawman/blob/3a179ded881db71d614a9c4b3c617b139e16d163/test/junit/src/test/scala/strawman/collection/immutable/LazyListTest.scala#L138)

Reply:
@dhanesharole Lukas's sample solves this case too.